### DISCUSSION TITLE
🎧 Kaizen 14 - There goes my PgHero - Apr. 5, 2024

### DISCUSSION THREAD

**Discussion by gerhard** [thechangelog/changelog.com#499] [Kaizen]
![shields.io](https://img.shields.io/badge/Last%20updated%20on%20May%207%2C%202024-PUBLISHED-blue?style=for-the-badge)

[🎧 This is now available as changelog.com/friends/38](https://changelog.com/friends/38)

<a href="https://changelog.com/friends/38" target="_blank"><img width="1440" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/55752b54-3567-43a5-ae52-104863a1caca"></a>

See all improvements in this 🐙 GitHub repository: [Kaizen 13 ... 14](https://github.com/thechangelog/changelog.com/compare/kaizen.13...kaizen.14)

---

This contains all things related to the next Kaizen episode.

Share your thoughts via comments below:
- Is there something in changelog.com that you, our dear listener, would like us to tackle?
- Is there something that **you** would like to help out with?
- Anything that we - @adamstac @jerodsanto - want to improve?

A rough timeline of what to expect (red line is today):

```mermaid
gantt
    title 🎧 Kaizen 14 timeline
    dateFormat  YYYY-MM-DD
    axisFormat %b %e, %Y

    section CONTRIBUTE
    💻 ALL CONTRIBUTIONS WELCOME :2024-01-12, 77d

    section RECORD

    🎙 Record :milestone, 2024-03-29
    ✂️ Edit :2024-03-29, 7d
    🚀 SHIP IT! :milestone, 2024-04-05

    section NEXT
    Kick-off Kaizen 15 :2024-04-05, 7d
```

---

### What changed? [Kaizen.13 ... Kaizen.14](https://github.com/thechangelog/changelog.com/compare/kaizen.13...master)
- Did we build a CDN?
  - 🤔 #501 
  - 😔 #486
  - 🙂 https://bunny.changelog.com cc @lawik 
    - [Changelog on Fastly](https://app.hyperping.io/report/mon_qVm5g1PlusuOP5)
    - [Changelog on Fly](https://app.hyperping.io/report/mon_DiW4YRTnpeksbW)
    - [Changelog on Bunny](https://app.hyperping.io/report/mon_wUgaULN1tV7Eqh)
    - [Changelog on Cloudflare](https://github.com/thechangelog/changelog.com/issues/486#issuecomment-1963019736)
  - 😔 #506 
- What about Neon.tech Postgres?
  - [Shout-out](https://github.com/thechangelog/changelog.com/pull/492#issuecomment-1885586669) @brendan-stephens from Neon.tech Support
  - 🙂 #507
  - 🤩 #508 
    - Shout-out to @kislerdm for https://github.com/kislerdm/neon-sdk-go
    - Is it enough?
    - What would make it better?
    - [Easily Anonymize Production Data in Postgres](https://neon.tech/blog/easily-anonymize-production-data-in-postgres)
- `55` commits later & `0` PRs
- Make it Work

---

### What else is coming up?

- Daggerize `action-slack-notify`
  - https://github.com/gerhard/daggerverse/tree/main/notify
- The CDN saga&trade;
- Daggerball on K8s
  - Functional CI - https://daggerverse.dev
  - New UI
  - One more thing...

---

### This is a follow-up to
- https://github.com/thechangelog/changelog.com/discussions/485
- https://github.com/thechangelog/changelog.com/discussions/480
- https://github.com/thechangelog/changelog.com/discussions/469

**Comment by gerhard** [comment]
`docker run -it -e DATABASE_URL="$(op read op://changelog/neon/url --account changelog.1password.com --cache)" -p 8081:8080 ankane/pghero`

<img width="1280" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/e41157a2-2199-4049-9fc3-f4988b7d5d1b">

<img width="1280" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/4e1f6f9e-674b-4ea3-9ba9-a0930184e79a">

<img width="1280" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/3a15cfad-8fc8-41c9-9e79-5af96caac819">

**Comment by gerhard** [comment]
If anyone wants to compare our 3 setups part of the CDN Saga™️ series, you can do so via:
1. https://changelog.com/ via Fastly
2. https://changelog-2024-01-12.fly.dev/ via Fly.io
3. https://changelog-2024-01-12.b-cdn.net via Bunny CDN

These are the response results for Mar 8 - Apr 7, 2024 according to https://hyperping.io

## 1. https://changelog.com/ via Fastly - `392ms`
<img width="1440" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/9a258bb0-9359-4792-bcd7-e9eb3ed65b30">

Checking the headers by running `httpstat` locally:
<img width="1433" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/e1928e89-6ac4-445e-8fb2-ed13b26c2d09">

## 2. https://changelog-2024-01-12.fly.dev/ via Fly.io - `265ms`
<img width="1440" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/34e7f1e8-4dc7-448e-bf7b-f6a6af732c32">

Checking the headers by running `httpstat` locally:
<img width="1425" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/5fc664fc-627a-46b8-a044-6b105ab3f0d5">

## 3. https://changelog-2024-01-12.b-cdn.net via Bunny CDN - `59ms`
<img width="1440" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/c6478452-320f-4f93-8039-608f0f71a755">

Checking the headers by running `httpstat` locally:
<img width="1433" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/156e0d2a-629e-4b78-9302-2863ab4ccdb1">

---
FTR: https://www.cdnperf.com/
<img width="1438" alt="image" src="https://github.com/thechangelog/changelog.com/assets/3342/18f93a81-2157-426d-bb49-c3410352ad1a">

**Comment by lawik** [comment]
But since you are running on Fly you have to test their object storage offering with built-in CDN-style caching. Tigris (who I am currently working with so enormously biased) seem like a great fit for this. Push your stuff with eager caching if you want immediate low latency, otherwise you get the usual second-fetch being fast.

**Reply by gerhard** [reply]
This keeps coming up, I am curious to learn more:
1. Am I right in assuming that this is a Cloudflare R2 alternative?
2. Do they have a pull-through cache, similar to how Bunny operates?
3. Does it support HTTP rules such as redirects, rewrites, etc.?

**Reply by adamstac** [reply]
I mentioned them because Lars intro'd me to them. The frequency bias is still one layer, not two.

**Reply by lawik** [reply]
Yeah, wrote this before I listened.

1. if you want geo-replication with R2 you better enjoy writing edge worker Cache API code because R2 doesn't do it by default. The marketing says yes. Threads on forum says no.
2. Yes, for buckets. Also can be actively told to cache eagerly which is rad. Not for your pages though unless you render them out to disk.
3. No. Not that I am aware.

It essentially rounds out the Fly offering of putting your apps close to your users by also putting your files close to your users.

if you ran more app instances in a geo-distributed fashion and had some light caching in-app on the heavily used pages you'd probably get good results.

But most likely you'll find one or two pathological cases where moving your app and DB too far apart makes something awfully slow.

**Reply by lawik** [reply]
Also, cool to be a bias with frequency. Signal all on my own :D

**Reply by gerhard** [reply]
Just to double-check my understanding:
1. Yes, with extras (e.g. object geo-replication that just works)
2. Not for files that do not exist in the bucket. Caching is configured on the object store, not on origins.
3. 👍

Based on the above, this still sounds like an R2 alternative with the added benefit that we could serve those files directly from the object store cache which can be configured to replicate closer to end-users. Tigris doesn't behave as a CDN in the sense that we can't declare origins, request rules, etc.

**Reply by lawik** [reply]
Yeah, it is more of an R2 alternative than a CDN in that regard. It will CDN your media files and all that but unless you render out your pages to file on change it will not solve your landing page etc.
